camayoc/tests/remote/rho/test_scan.py

```diff
@@ -13,35 +13,239 @@
+            auth_add(
+                {
+                    'name': auth['name'],
+                    'username': auth['username'],
+                    'password': None,
+                },
+                [(
+                    CONNECTION_PASSWORD_INPUT,
+                    auth['password']
+                )],
+            )
+    chunk_size = 5
+    chunks = [vms[i:i+chunk_size] for i in range(0, len(vms), chunk_size)]
+    for chunk in chunks:
+        machines_to_wait = []
+        for vm in chunk:
+            if vm.runtime.powerState == 'poweredOff':
+                vm.PowerOnVM_Task()
+                machines_to_wait.append(inventory[vm.name])
+        wait_until_live(machines_to_wait)
+        for vm in chunk:
+            if vm.runtime.powerState == 'poweredOn':
+                machine = inventory[vm.name]
+                for auth in auths:
+                    profile_name = machine['hostname'] + '-' + auth['name']
+                    try:
+                        result = scan_machine(machine, auth, profile_name)
+                    except (AssertionError, pexpect.exceptions.EOF) as err:
+                        result = None
+                    SCAN_RESULTS[profile_name] = result
+
+                vm.PowerOffVM_Task()
+
+
+def scan_permutations():
+    """Generate a tuple of hostname and auth to be used on scans."""
+    try:
+        config = get_config()
+        auths = config['rho']['auths']
+        inventory = config['inventory']
+        return list(itertools.product(inventory, auths))
+    except (ConfigFileNotFoundError, KeyError):
+        return []
+
+
+def get_permutation_id(param):
+    """Return a string representation for each  :func:`test_scan` parameter."""
+    return param.get('name') or param.get('hostname')
+
+
+@pytest.mark.parametrize(
+    'machine,auth', scan_permutations(), ids=get_permutation_id)
+def test_scan(machine, auth):
+    """Test each scan report.
 
     :id: d9eb29bd-1b61-421a-b680-f494e868b11e
-    :description: Test the facts from a scan report
-    :steps: For each fact collected, assert it is as expected
-    :expectedresults:
-        The fact will contain expected data based on privilege level of
-        credentials used to acquire it.
+    :description: Test if the scan produced the expected report for a given
+        auth and machine.
+    :steps: For each machine in the inventory perform one scan per available
+        auth.
+    :expectedresults: The generate report must have the expected fact values.
     """
+    result = SCAN_RESULTS[machine['hostname'] + '-' + auth['name']]
+    if result is None:
+        pytest.fail(
+            'Was not able to scan {0[hostname]} using the auth {1[name]}.\n'
+            'Machine object {0}\n\nAuth object: {1}'
+            .format(machine, auth)
+        )
+
+    with open(result) as f:
+        fieldnames = csv.DictReader(f).fieldnames
+    assert sorted(fieldnames) == sorted(RHO_ALL_FACTS)
+
+
+    return
```

Comment:
Returning earlier for now, once fact values assertions are in place this will be dropped.

Reply:
So are you are not thinking about parametrizing a test on the facts in the scan report? Rather, are we going back to just having one test result per scan?

Reply:
What I have in mind is that we can have one failure and many fact values on it. That will enable us to identify what is the machine and auth pair that is problematic and all facts that doesn't have the expected value by looking at a single failure. What do you think?

Check https://github.com/quipucords/camayoc/pull/88#discussion_r155223740 for more details and decisions I made to have this implementation.

Reply:
I see so all the facts values will be tested and all the failures lumped together into one failure result and message.

I like that, while it was impressive to see 4,000 tests running, the results were not easy to interpret.